keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -0,0 +1,494 @@
+
+### Logging Interface
+ 
+When choosing new structured logging API following criteria has been used:
+* Selected API should provide abstract structured logging interface so that it would be easy to inject different implementations for the same API.
+* Selected API should be minimalistic so that it should be easy to reimplement, wrap or to mock it.
+* Applicability of the selected API should not be limited to the Kubernetes control plane but it should be possible to use the same API in other components running on Kubernetes clusters which are also implemented in go.
+* Selected API should support structured logging by providing a way to define a set of key-value pairs for each log entry.
+* There should at least one production ready implementation of the selected API which could produce JSON formatted logs.
+* It should be possible to produce klog compatible logs from the selected API which could be used during the migration to a new API.
+
+Based on this criteria https://github.com/go-logr/logr log interface has been selected as a logging API for Kubernetes.
+
+Apart from meeting the criteria listed above logr has a few nice features:
+* Constant log message together with logger name provide stable identification for log entries which greatly simplifies querying logs.
+* Limiting log levels to just Info and Error makes it much easier to choose the proper log level for given log.
+* Using key-values pairs for defining log fields allows more efficient implementation of the logger.
+
+#### Logger initialization
+
+Logger implementation initialization should be done at startup of component and passed down the stack as an interface. Aside of initialization of logger, there should not be any dependency on specific implementation of logger Interface. 
+
+#### Storing Logger and metadata in Context
+
+We would like to propose that logger interface and all metadata collected trough call stack to be stored in Golang Context object. This way creating one canonical way to pass contextual information across the call stack. Some pros of this approach:
+* Standard method in Golang to pass scoped variables across API boundaries and between processes.
+* No need for additional function argument for passing logging API through call stack
+* Context is used similarly by many frameworks including for instance OpenCensus tracing API.
+* Accessing metadata will be independent from Logging and could be reused by other functionalities.
+
+Access to context variables should be done though predefined functions. For that we propose to a new module to kubernetes component base called `kontext`. This new module will include all utility functions for storing and retrieving logger and metadata needed by it. 
```

Comment:
This does look like something that would also be useful outside of Kubernetes. A new `k8s.io/kontext` module will of course also work for that, but why not go all the way and implement it as a package in `github.com/go-logr/logr/lcontext`?

That'll make it more discoverable for people who find `logr` outside of Kubernetes.

Reply:
I agree that this is a good idea. I already discussed it with @thockin and @DirectXMan12 that `logr` itself should interact with context.

Reply:
I also think it is useful, but it also comes with the risks that we already discussed here (logger not set, for example). In fact, Dave Cheney advocated to not do it: https://dave.cheney.net/2017/01/26/context-is-for-cancelation

I would still prefer to use context to pass a logger if we can come up with a helper package that deals with the situation when the context doesn't have a logger, simply because it eases the migration to that new logger API.

If we require every component to change its API and thus do a major version bump (whether it is because of adding a new parameter to functions or implicitly by changing how the component is to be used), then I fear that it will not get adopted and developers will continue to import and use the current zoo of logging implementations.

Reply:
Apart from introducing a structure to Kubernetes logs within this KEP we would like also to make logs more contextul, meaning that logs apart from normal data could also contain information about the context in which given log entry has been produced like for instance references to pods or nodes related to a given log entry.

This by itself indicates very close relationship between the logger and the context but it does not automatically mean that we should pass logger within a context.

When it comes to the implementation of contextual logging this can be approached in a few different ways:

- you can have a stateless logger, for instance injected during the library initialization, which gets context as an argument to its methods and extracts objects to be logged from this context,
- you can have a stateful logger directly passed with each call,
- you can have a stateful logger indirectly passed through context.

First solution does not scale very well with the number of different objects which could be logged as a log entry context because logger needs to know when to look for those.

Second solution is very disruptive to APIs of libraries which log anything and it is hard to be implemented in a large system like Kubernetes.

Third has some problems like for instance the problem with handling situation when there is no logger in the context but taking all into account it seems to be the most pragmatic one.

Reply:
What about offering a settable singleton logger? Not that singletons are necessarily a great pattern either, but it also avoids changing APIs and passing the logger through a Context.

Even if we're passing Context around, some functions APIs may need to change to accept a Context. The current glog/klog logger is effectively a singleton logger.

Reply:
Thanks for pointing this out. Context-ful logs and integration with Tracing will require more discussion and should be tackled by a dedicated KEP.
We have decided to improve klog, which is Singleton logger.